core/arch/arm/plat-imx/imx_caam.h

```diff
@@ -0,0 +1,40 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2019 Bryan O'Donoghue
+ *
+ * Bryan O'Donoghue <bryan.odonoghue@linaro.org>
+ */
+
+#ifndef __IMX_CAAM_H__
+#define __IMX_CAAM_H__
+
+#include <imx-regs.h>
+#include <stdint.h>
+
+struct imx_caam_job_ring {
+	uint32_t			jrmidr_ms;
+	uint32_t			jrmidr_ls;
+};
+
+#if defined(CFG_MX6ULL)
+#define CAAM_NUM_JOB_RINGS		0
+#else
+#define CAAM_NUM_JOB_RINGS		4
```

Comment:
At least the Security Reference Manuals I have for i.MX6Q and i.MX6UL document different numbers of job rings. Maybe @MrVan can clarify whether it is okay to assume four jobrings for all devices.

Reply:
ah damn.
I _thought_ I probably wasn't getting full coverage on job-ring numbers - let me review the two you have flagged - thanks.

Reply:
Actually I believe four is the correct number.

https://github.com/u-boot/u-boot/blob/master/include/fsl_sec.h - line 84
https://github.com/u-boot/u-boot/blob/master/drivers/crypto/fsl/jr.c - line 636

The address-space maps 4 job-ring queues. You're right that not all processors implement 4 queues but the address space offset is valid.

It's up to the CAAM drivers - and their DTS configs to describe which _runtime_ job-rings are used.

4 is the right value here though.